File: internal/db/models/jobs.go
package models

import (
	"encoding/json"
	"time"

	"github.com/uptrace/bun"
)

// ─── Job ─────────────────────────────────────────────────────────────────────

// Job type constants.
const (
	JobTypeSync            = "sync"
	JobTypeImport          = "import"
	JobTypeExport          = "export"
	JobTypeMetadataRefresh = "metadata_refresh"
)

// Job source constants.
const (
	JobSourceSteam     = "steam"
	JobSourceEpic      = "epic"
	JobSourcePSN       = "psn"
	JobSourceGOG       = "gog"
	JobSourceManual    = "manual"
	JobSourceNexorious = "nexorious"
	JobSourceCSV       = "csv"
	JobSourceSystem    = "system"
)

// Job status constants.
const (
	JobStatusPending             = "pending"
	JobStatusProcessing          = "processing"
	JobStatusCompleted           = "completed"
	JobStatusFailed              = "failed"
	JobStatusCancelled           = "cancelled"
	JobStatusCompletedWithErrors = "completed_with_errors"
)

// Job priority constants.
const (
	JobPriorityLow    = "low"
	JobPriorityNormal = "normal"
	JobPriorityHigh   = "high"
)

type Job struct {
	bun.BaseModel `bun:"table:jobs"`

	ID            string     `bun:"id,pk"                    json:"id"`
	UserID        string     `bun:"user_id,notnull"          json:"user_id"`
	JobType       string     `bun:"job_type,notnull"         json:"job_type"`
	Source        string     `bun:"source,notnull"           json:"source"`
	Status        string     `bun:"status,notnull"           json:"status"`
	Priority      string     `bun:"priority,notnull"         json:"priority"`
	FilePath      *string    `bun:"file_path"                json:"file_path"`
	TotalItems    int        `bun:"total_items,notnull"      json:"total_items"`
	ErrorMessage  *string    `bun:"error_message"            json:"error_message"`
	AutoRetryDone bool       `bun:"auto_retry_done,notnull"  json:"auto_retry_done"`
	CreatedAt     time.Time  `bun:"created_at,notnull"       json:"created_at"`
	StartedAt     *time.Time `bun:"started_at"               json:"started_at"`
	CompletedAt   *time.Time `bun:"completed_at"             json:"completed_at"`
}

// IsActive returns true if the job is pending or processing.
func (j *Job) IsActive() bool {
	return j.Status == JobStatusPending || j.Status == JobStatusProcessing
}

// IsTerminal returns true if the job has reached a terminal state.
func (j *Job) IsTerminal() bool {
	return j.Status == JobStatusCompleted ||
		j.Status == JobStatusFailed ||
		j.Status == JobStatusCancelled ||
		j.Status == JobStatusCompletedWithErrors
}

// DurationSeconds returns elapsed seconds from StartedAt to CompletedAt
// (or now if still running). Returns nil if StartedAt is not set.
func (j *Job) DurationSeconds() *float64 {
	if j.StartedAt == nil {
		return nil
	}
	end := time.Now()
	if j.CompletedAt != nil {
		end = *j.CompletedAt
	}
	d := end.Sub(*j.StartedAt).Seconds()
	return &d
}

// ─── JobItem ─────────────────────────────────────────────────────────────────

// JobItem status constants.
const (
	JobItemStatusPending       = "pending"
	JobItemStatusProcessing    = "processing"
	JobItemStatusCompleted     = "completed"
	JobItemStatusPendingReview = "pending_review"
	JobItemStatusSkipped       = "skipped"
	JobItemStatusFailed        = "failed"
	JobItemStatusIGDBFailed    = "igdb_failed"
)

type JobItem struct {
	bun.BaseModel `bun:"table:job_items"`

	ID              string          `bun:"id,pk"                    json:"id"`
	JobID           string          `bun:"job_id,notnull"           json:"job_id"`
	UserID          string          `bun:"user_id,notnull"          json:"user_id"`
	ItemKey         string          `bun:"item_key,notnull"         json:"item_key"`
	SourceTitle     string          `bun:"source_title,notnull"     json:"source_title"`
	SourceMetadata  json.RawMessage `bun:"source_metadata,notnull"  json:"source_metadata"`
	Status          string          `bun:"status,notnull"           json:"status"`
	Result          json.RawMessage `bun:"result,notnull"           json:"result"`
	ErrorMessage    *string         `bun:"error_message"            json:"error_message"`
	IGDBCandidates  json.RawMessage `bun:"igdb_candidates,notnull"  json:"igdb_candidates"`
	ResolvedIGDBID  *int            `bun:"resolved_igdb_id"         json:"resolved_igdb_id"`
	MatchConfidence *float64        `bun:"match_confidence"         json:"match_confidence"`
	CreatedAt       time.Time       `bun:"created_at,notnull"       json:"created_at"`
	ProcessedAt     *time.Time      `bun:"processed_at"             json:"processed_at"`
	ResolvedAt      *time.Time      `bun:"resolved_at"              json:"resolved_at"`
}

File: ui/frontend/src/types/jobs.ts
/**
 * Types for background job management.
 * Jobs include sync, import, and export operations.
 */

// ============================================================================
// Enums
// ============================================================================

export enum JobType {
  SYNC = 'sync',
  IMPORT = 'import',
  EXPORT = 'export',
  METADATA_REFRESH = 'metadata_refresh',
}

export enum JobSource {
  STEAM = 'steam',
  EPIC = 'epic',
  PSN = 'psn',
  GOG = 'gog',
  HUMBLE_BUNDLE = 'humble-bundle',
  MANUAL = 'manual',
  DARKADIA = 'darkadia',
  NEXORIOUS = 'nexorious',
  CSV = 'csv',
  SYSTEM = 'system',
}

export enum JobStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

export enum JobItemStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  PENDING_REVIEW = 'pending_review',
  SKIPPED = 'skipped',
  FAILED = 'failed',
}

export type JobPriority = 'low' | 'normal' | 'high';

// ============================================================================
// Interfaces
// ============================================================================

/**
 * Progress counts by JobItem status.
 */
export interface JobProgress {
  pending: number;
  processing: number;
  completed: number;
  pendingReview: number;
  skipped: number;
  failed: number;
  total: number;
  percent: number;
}

export interface Job {
  id: string;
  userId: string;
  jobType: JobType;
  source: JobSource;
  status: JobStatus;
  priority: JobPriority;
  progress: JobProgress;
  totalItems: number;
  errorMessage: string | null;
  filePath: string | null;
  createdAt: string;
  startedAt: string | null;
  completedAt: string | null;
  isTerminal: boolean;
  durationSeconds: number | null;
}

export interface JobTypeStatus {
  isActive: boolean;
  activeJobId: string | null;
  lastCompletedJobId: string | null;
  lastCompletedAt: string | null;
}

export interface JobFilters {
  jobType?: JobType;
  source?: JobSource;
  status?: JobStatus;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
}

// Note: JobChildrenFilters removed - child jobs no longer supported

export interface JobListResponse {
  jobs: Job[];
  total: number;
  page: number;
  perPage: number;
  pages: number;
}

export interface JobItem {
  id: string;
  jobId: string;
  itemKey: string;
  sourceTitle: string;
  status: JobItemStatus;
  errorMessage: string | null;
  resultGameTitle: string | null;
  resultIgdbId: number | null;
  resultUserGameId: string | null;
  createdAt: string;
  processedAt: string | null;
  igdbCandidatesJson?: string; // Optional - present for PENDING_REVIEW items
  externalGameId: string | null;
}

export interface JobItemListResponse {
  items: JobItem[];
  total: number;
  page: number;
  pageSize: number;
  pages: number;
}

export interface JobCancelResponse {
  success: boolean;
  message: string;
  job: Job | null;
}

export interface JobDeleteResponse {
  success: boolean;
  message: string;
  deletedJobId: string;
}

export interface JobsSummary {
  runningCount: number;
  failedCount: number;
}

export interface PendingReviewCountResponse {
  pendingReviewCount: number;
  countsBySource: Record<string, number>;
}

export interface RetryFailedResponse {
  success: boolean;
  message: string;
  retriedCount: number;
}

export interface JobItemDetail extends JobItem {
  sourceMetadataJson: string;
  resultJson: string;
  igdbCandidatesJson: string;
  resolvedIgdbId: number | null;
  resolvedAt: string | null;
}

export interface SyncChangeItem {
  title: string;
  oldStatus?: string | null;
  newStatus?: string | null;
}

export interface RecentJobDetail {
  id: string;
  jobType: JobType;
  source: JobSource;
  status: string;
  createdAt: string;
  completedAt: string | null;
  errorMessage: string | null;
  totalItems: number;
  completedCount: number;
  skippedCount: number;
  failedCount: number;
  progress: JobProgress;
  addedItems: SyncChangeItem[];
  updatedItems: SyncChangeItem[];
  removedItems: SyncChangeItem[];
  statusChangedItems: SyncChangeItem[];
  skippedItems: SyncChangeItem[];
  alreadyInLibraryItems: SyncChangeItem[];
}

export interface RecentJobsResponse {
  jobs: RecentJobDetail[];
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Get a human-readable label for a job type.
 */
export function getJobTypeLabel(type: JobType): string {
  const labels: Record<JobType, string> = {
    [JobType.SYNC]: 'Sync',
    [JobType.IMPORT]: 'Import',
    [JobType.EXPORT]: 'Export',
    [JobType.METADATA_REFRESH]: 'Metadata Refresh',
  };
  return labels[type] || type;
}

/**
 * Get a human-readable label for a job source.
 */
export function getJobSourceLabel(source: JobSource): string {
  const labels: Record<JobSource, string> = {
    [JobSource.STEAM]: 'Steam',
    [JobSource.EPIC]: 'Epic Games',
    [JobSource.PSN]: 'PlayStation Network',
    [JobSource.GOG]: 'GOG',
    [JobSource.HUMBLE_BUNDLE]: 'Humble Bundle',
    [JobSource.MANUAL]: 'Manual',
    [JobSource.DARKADIA]: 'Darkadia',
    [JobSource.NEXORIOUS]: 'Nexorious',
    [JobSource.CSV]: 'CSV',
    [JobSource.SYSTEM]: 'System',
  };
  return labels[source] || source;
}

/**
 * Get a human-readable label for a job status.
 */
export function getJobStatusLabel(status: JobStatus): string {
  const labels: Record<JobStatus, string> = {
    [JobStatus.PENDING]: 'Pending',
    [JobStatus.PROCESSING]: 'Processing',
    [JobStatus.COMPLETED]: 'Completed',
    [JobStatus.FAILED]: 'Failed',
    [JobStatus.CANCELLED]: 'Cancelled',
  };
  return labels[status] || status;
}

/**
 * Get CSS classes for a job status badge.
 */
export function getJobStatusVariant(
  status: JobStatus,
): 'default' | 'secondary' | 'destructive' | 'outline' {
  switch (status) {
    case JobStatus.COMPLETED:
      return 'default';
    case JobStatus.PROCESSING:
      return 'secondary';
    case JobStatus.FAILED:
      return 'destructive';
    case JobStatus.PENDING:
    case JobStatus.CANCELLED:
    default:
      return 'outline';
  }
}

/**
 * Format duration in seconds to human-readable string.
 */
export function formatDuration(seconds: number | null): string {
  if (seconds === null || seconds === undefined) return '-';

  if (seconds < 60) {
    return `${Math.round(seconds)}s`;
  } else if (seconds < 3600) {
    const minutes = Math.floor(seconds / 60);
    const secs = Math.round(seconds % 60);
    return secs > 0 ? `${minutes}m ${secs}s` : `${minutes}m`;
  } else {
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.round((seconds % 3600) / 60);
    return minutes > 0 ? `${hours}h ${minutes}m` : `${hours}h`;
  }
}

/**
 * Format a date string to a relative time (e.g., "5m ago"). `nullLabel` is the
 * placeholder returned for a missing date (e.g. '-' for activity, 'Never' for
 * last-sync timestamps).
 */
export function formatRelativeTime(dateStr: string | null, nullLabel = '-'): string {
  if (!dateStr) return nullLabel;
  const date = new Date(dateStr);
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  const diffMins = Math.floor(diffMs / 60000);
  const diffHours = Math.floor(diffMins / 60);
  const diffDays = Math.floor(diffHours / 24);

  if (diffMins < 1) return 'Just now';
  if (diffMins < 60) return `${diffMins}m ago`;
  if (diffHours < 24) return `${diffHours}h ago`;
  if (diffDays < 7) return `${diffDays}d ago`;
  return date.toLocaleDateString();
}

/**
 * Check if a job is currently in progress (not terminal).
 */
export function isJobInProgress(job: Job): boolean {
  return !job.isTerminal;
}

/**
 * Check if a job can be cancelled.
 * Jobs can be cancelled at any point before they reach a terminal state.
 */
export function canCancelJob(job: Job): boolean {
  return !job.isTerminal;
}

/**
 * Check if a job can be deleted.
 * Jobs can be deleted at any point (terminal or not).
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function canDeleteJob(_job: Job): boolean {
  return true;
}

/**
 * Get a human-readable label for a job item status.
 */
export function getJobItemStatusLabel(status: JobItemStatus): string {
  const labels: Record<JobItemStatus, string> = {
    [JobItemStatus.PENDING]: 'Pending',
    [JobItemStatus.PROCESSING]: 'Processing',
    [JobItemStatus.COMPLETED]: 'Completed',
    [JobItemStatus.PENDING_REVIEW]: 'Needs Review',
    [JobItemStatus.SKIPPED]: 'Skipped',
    [JobItemStatus.FAILED]: 'Failed',
  };
  return labels[status];
}

/**
 * Get CSS classes for a job item status badge.
 */
export function getJobItemStatusVariant(
  status: JobItemStatus,
): 'default' | 'secondary' | 'destructive' | 'outline' {
  switch (status) {
    case JobItemStatus.COMPLETED:
      return 'default';
    case JobItemStatus.FAILED:
      return 'destructive';
    case JobItemStatus.PENDING_REVIEW:
      return 'secondary';
    default:
      return 'outline';
  }
}
